feat(StatusQ.Popups) Adding SearchPopup component

alexandraB99 opened the pull request:
Closes #264

PascalPrecht commented:
Haven't reviewed the code yet, but wanted to try it out. Unfortunately, I'm unable to try your changes in the demo application because this commit seems to break the sandbox app when navigating to the demo app.

All I'm getting is

```
zsh: segmentation fault  ./bin/sandbox
```

Are you experiencing the same?

alexandraB99 commented:
> Haven't reviewed the code yet, but wanted to try it out. Unfortunately, I'm unable to try your changes in the demo application because this commit seems to break the sandbox app when navigating to the demo app.
> 
> All I'm getting is
> 
> ```
> zsh: segmentation fault  ./bin/sandbox
> ```
> 
> Are you experiencing the same?

@PascalPrecht no. I just rebased and tried with a fresh build and still I see no issues.

PascalPrecht commented on src/StatusQ/Controls/StatusBaseButton.qml:
```diff
@@ -16,35 +24,26 @@ Rectangle {
     }
 
-    property StatusIconSettings icon: StatusIconSettings {
+    property StatusIconSettings iconSettings: StatusIconSettings {
```
This and the move to `Button` introduces a breaking change which we should avoid for now. I could see that this was probably done to be consistent with how icon settings work with `StatusMenuItem`. There however, we had to use `iconSettings` over `icon` because the built-in `icon` property couldn't be mutated.

I'm very fine with moving to `iconSettings` altogether in the long run but doubt that this breaking change is needed for this PR.

Maybe you can shine some light onto this :D I'm probably missing something

alexandraB99 replied:
see above regarding the background

PascalPrecht commented on src/StatusQ/Popups/StatusPopupMenu.qml:
```diff
@@ -19,6 +19,8 @@ Menu {
     property var openHandler
     property var closeHandler
+    property string currentImg
+    property string currentIdColor
```
It doesn't seem to be obvious why these ones are around.  Is this fixing some bug?

alexandraB99 replied:
no. those are exposed so that it can be set in the Button that opens the popup menu.

PascalPrecht replied:
Got it!

We should come up with better names though. Especially for `currentIdColor`. How about `currentItemColor` and then we can also go with `currentItemImageSource`?

alexandraB99 replied:
changed to `currentOptionItemColor` and `currentOptionItemImgSource`

PascalPrecht replied:
Great

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,192 @@
+                    input.bgVisible: false
+                }
+            }
+            StatusMenuSeparator { Layout.fillWidth: true }
```
Semantically, this is not a menu, so we might want to consider introducing a general `StatusSeparator` in the near future.

alexandraB99 replied:
Can we just rename this one if there is no difference between those 2? Maybe create a task for the near future?

PascalPrecht replied:
Yea absolutely. Let's just keep this as is and rename this in the future.

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,192 @@
+                                visible: ((root.selectionBadgeImage === "") && (root.selectionBadgePrimaryText !== "Anywhere"))
+                                color: root.selectionBadgeIdenticonColor
+                                name: root.selectionBadgePrimaryText
+                            }
```
> We're using this mechanism where we first try StatusRoundedImage and fallback to LetterIdenticon if it fails (or go to letter identicon straight away if no image source was provided. Maybe we can use a similar logic here?

> Here's an example of this: https://github.com/status-im/StatusQ/blob/master/src/StatusQ/Components/StatusChatListItem.qml#L59-L103

Maybe this applies here as well?

alexandraB99 replied:
please see above

PascalPrecht replied:
Not really sure what you're referring to here...

alexandraB99 replied:
what do you mean?

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,192 @@
+                        statusListItemSubTitle.height: model.content !== "" ? 20 : 0
+                        statusListItemSubTitle.elide: Text.ElideRight
+                        statusListItemSubTitle.color: Theme.palette.black
+                        titleAsideText: model.time //Utils.formatTime(timestamp)
```
Left over comment?

alexandraB99 replied:
removed

PascalPrecht reviewed (changes requested):
Massive work! As commented above, I couldn't play around with this yet, so I'll give it another spin once the app crash is resolved.

PascalPrecht commented:
> @PascalPrecht no. I just rebased and tried with a fresh build and still I see no issues.

@alexandraB99 that's interesting. Did you run `sandbox/scripts/build` before `./bin/sandbox` ?
If not, can you give it another spin after running the build script?  I can't access the demo application without sandbox crashing in this commit...

PascalPrecht commented:
Sorry, to clarify: the demo app can be opened but opening the community in the demo app crashes. That's what I'm referring to.

borismelnik commented on src/StatusQ/Components/StatusListItem.qml:
```diff
@@ -1,4 +1,6 @@
 import QtQuick 2.13
+import QtQuick.Controls 2.12
+import QtQuick.Layouts 1.12
```
If you do't mind pls use minor version of QtQuick import

alexandraB99 replied:
You're referring to Controls and Layouts? because QtQuick was there already

borismelnik commented on src/StatusQ/Components/StatusListItem.qml:
```diff
@@ -198,12 +218,69 @@ Rectangle {
                 wrapMode: Text.WrapAtWordBoundaryOrAnywhere
             }
+
+            Control {
```
and i think if use some customization we should create file for it `StatusListItemBadge`

alexandraB99 replied:
done

borismelnik commented on src/StatusQ/Controls/StatusBaseButton.qml:
```diff
@@ -1,10 +1,18 @@
 import QtQuick 2.14
+import QtQuick.Controls 2.12
```
The same as before. Qt creator sometimes dont understand which modules is default.

alexandraB99 replied:
please see above, you mean QtQuick or QtQuick.Controls import?

borismelnik commented on src/StatusQ/Controls/StatusBaseButton.qml:
```diff
@@ -1,10 +1,18 @@
 import StatusQ.Components 0.1
 
-Rectangle {
+Button {
```
i agreed. Why `Button`?

alexandraB99 replied:
Because I want to re-use the background, which is the same and all other features that might be needed and so I don't like to duplicate code.

borismelnik commented on src/StatusQ/Controls/StatusBaseInput.qml:
```diff
@@ -51,27 +52,29 @@ Item {
     implicitHeight: multiline ? Math.max(edit.implicitHeight + topPadding + bottomPadding, 44) : 44
 
-    Rectangle {
+    Item {
```
Actually i dont understand why you create two components instead of one?

alexandraB99 replied:
What do you mean?

PascalPrecht replied:
I think the reason this was done, was so that the `Rectangle` could be made invisible while keeping the element around.

alexandraB99 replied:
In this case isn't it better to use an Item and set its visibility to false instead? Rectangles are expensive when it comes to performance (due to gl calls) since they suppose to paint on screen and it's recommendable to use only when we need to really paint (eg have some background color)

PascalPrecht replied:
I was just trying to interpret what this change from `Rectangle` to `Item` + the introduction of another `Rectangle` below was about :D

alexandraB99 replied:
Probably some clipping issue, removed now as I switched to use a TextInput instead

PascalPrecht replied:
Cool!

borismelnik commented on src/StatusQ/Controls/StatusBaseInput.qml:
```diff
@@ -179,10 +182,10 @@ Item {
+        visible: edit.text.length != 0 &&
+                 statusBaseInput.clearable &&
+                 !statusBaseInput.multiline &&
+                 edit.activeFocus
```
Its formatter stuff?

alexandraB99 replied:
Yes, ctrl+i

borismelnik commented on src/StatusQ/Popups/StatusModal.qml:
```diff
@@ -43,7 +43,7 @@ QC.Popup {
 
     contentItem: Column {
-        width: parent.width
+        width: parent ? parent.width : 0
```
sometimes parent can be null?

alexandraB99 replied:
was when using the contentItem, switched back to use content now

PascalPrecht commented:
@alexandraB99 Since you've already figured out, using `contentItem` in `StatusModal` breaks the application. `StatusModal` has a dedicated property for content items which is called `content`. You can see how it's used in other modal examples in the sandbox application.

Using that property, the app no longer breaks. However, it seems that the search component doesn't look quite how it's been designed (not sure why QtCreator shows something differeent but that's kinda what I've been referring to when I said it doesn't seem to be reliable):

<img width="809" alt="Screenshot 2021-07-30 at 14 13 33" src="https://user-images.githubusercontent.com/445106/127651686-32e39ed6-d10f-4fc2-a692-d480c5cf3a9e.png">


Also, notice that, when you want to access components that live inside `content`, they have to be accessed via `contentComponent` because the content is loaded via a `Loader`.

alexandraB99 commented on src/StatusQ/Controls/StatusBaseButton.qml:
```diff
@@ -16,35 +24,26 @@ Rectangle {
     property bool loading: false
-
-    property alias hovered: sensor.containsMouse
+    //property bool hovered: sensor.containsMouse
```
removed

alexandraB99 commented on src/StatusQ/Components/StatusListItem.qml:
```diff
@@ -198,12 +218,69 @@ Rectangle {
+                            visible: (statusListItem.badgeImage === "")
+                            color: statusListItem.badgeIdenticonColor
+                            name: statusListItem.badgePrimaryText
+                        }
```
do you mean to use `image.source.toString()` instead of `=== ""`?

PascalPrecht replied:
Oh I'm seeing this comment just now... So the mechanism I was referring to was:

1. First try and see if there's an `image.source`, then we know we want to render a `RoundedImage`.
2. Now the question is if it's an identicon or not. If yes, we need to render a border around it. 
3. If loading the image fails, we fallback to letter identicon
4. if there was no image.source in the first place, we probably deal with an icon and `isLetterIdenticon` true, in which case we'll render a letter identicon right away.

alexandraB99 replied:
not sure if in this case the user can set manually properties like `isLetterIdenticon` or `isIdenticon`. I assume this info will come from some model, plus I don't see any borders surrounding the images looking at the design, maybe I'm missing something?

PascalPrecht replied:
The moment the popup gets generated this information should be available. Basically, the rule is, if there's no profile image, but an image source is provided, then it's an identicon. Identicon profile images Always render with a border in the application. Need to check the design if it's missing there. You can see this being done in many other places in StatusQ

alexandraB99 commented on src/StatusQ/Components/StatusListItem.qml:
```diff
@@ -183,13 +191,25 @@ Rectangle {
+                font.pixelSize: 10
+                color: Theme.palette.baseColor1
+                visible: !!statusListItem.titleAsideText
+            }
```
done

alexandraB99 commented on src/StatusQ/Components/StatusListItem.qml:
```diff
@@ -33,8 +35,13 @@ Rectangle {
+    property string badgeImage: ""
+    property string badgePrimaryText: ""
+    property string badgeSecondaryText: ""
+    property string badgeIdenticonColor: ""
```
done

alexandraB99 commented on sandbox/DemoApp.qml:
```diff
@@ -564,12 +564,180 @@ Rectangle {
+                            ListElement { name: "MyCommunity"; type: "Communities"; time: ""; content: ""; badgeImage: ""; badgePrimaryText: ""; badgeSecondaryText: ""; badgeIdenticonColor: "" }
+                            ListElement { name: "Foo"; type: "Communities"; time: ""; content: ""; badgeImage: ""; badgePrimaryText: ""; badgeSecondaryText: ""; badgeIdenticonColor: "" }
+                        }
+                        property var searchResultsElse: ListModel {
```
done

alexandraB99 commented on sandbox/DemoApp.qml:
```diff
@@ -564,12 +564,180 @@ Rectangle {
+                        id: searchModelSimTimer
+                        interval: 500
+                        //dummy models
+                        property var searchResultsC: ListModel {
```
done

alexandraB99 commented:
> @alexandraB99 Since you've already figured out, using `contentItem` in `StatusModal` breaks the application. `StatusModal` has a dedicated property for content items which is called `content`. You can see how it's used in other modal examples in the sandbox application.
> 
> Using that property, the app no longer breaks. However, it seems that the search component doesn't look quite how it's been designed (not sure why QtCreator shows something differeent but that's kinda what I've been referring to when I said it doesn't seem to be reliable):
> 
> <img alt="Screenshot 2021-07-30 at 14 13 33" width="809" src="https://user-images.githubusercontent.com/445106/127651686-32e39ed6-d10f-4fc2-a692-d480c5cf3a9e.png">
> 
> Also, notice that, when you want to access components that live inside `content`, they have to be accessed via `contentComponent` because the content is loaded via a `Loader`.

switched to use `content`, crash is gone now, please verify

PascalPrecht commented on src/StatusQ/Components/StatusListItemBadge.qml:
```diff
@@ -0,0 +1,65 @@
+            }
+            StatusBaseText {
+                color: Theme.palette.baseColor1
+                text: root.badgePrimaryText
```
This should be `Font.Medium` weight

alexandraB99 replied:
done

PascalPrecht commented on src/StatusQ/Components/StatusListItemBadge.qml:
```diff
@@ -0,0 +1,65 @@
+            StatusBaseText {
+                color: Theme.palette.baseColor1
+                visible: !!root.badgeSecondaryText
+                text: root.badgeSecondaryText
```
This should be `Font.Medium` weight

alexandraB99 replied:
done

PascalPrecht commented on src/StatusQ/Controls/StatusFlatButton.qml:
```diff
@@ -14,7 +14,7 @@ StatusBaseButton {
     disabledTextColor: Theme.palette.baseColor1
 
-    border.color: type === StatusBaseButton.Type.Normal || hovered ? "transparent"
+    borderColor: type === StatusBaseButton.Type.Normal || hovered ? "transparent"
```
I assume this is a result of moving to `Button` type in `StatusBaseButton` as well?

alexandraB99 replied:
removed

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,198 @@
+                    input.placeholderText: "Search"
+                    input.bgVisible: false
+                }
+            }
```
The size of this icon seems wrong. This is what has been designed:

<img width="778" alt="Screenshot 2021-08-03 at 23 06 56" src="https://user-images.githubusercontent.com/445106/128086203-69a8f6da-5f4a-404f-ab93-68d808ba33f8.png">

This is what it looks like:

<img width="737" alt="Screenshot 2021-08-03 at 23 07 23" src="https://user-images.githubusercontent.com/445106/128086252-9e9054c9-adc2-4588-a370-35922b9e8040.png">


If `StatusBaseInput` is the issue here and not flexible enough, it's perfectly fine to introduce a custom inline input for this scenario.

alexandraB99 replied:
Exactly, does it really make sense to duplicate a whole text input for just making the icon bigger or maybe make the existing StatusInput able to have a bigger icon?

PascalPrecht replied:
I'm honestly happy fine either way. The search input is a super special version of an input. If it can't be easily derived from `StatusBaseInput` that's perfectly fine.

I personally would probably build a new one (that's also dramatically simpler than `StatusBaseInput`. But if you prefer making the base one more flexible, please go ahead!

alexandraB99 replied:
Done, switched to use a TextInput instead

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,198 @@
+                                implicitWidth: 16
+                                implicitHeight: 16
+                                visible: !!root.selectionBadgeImage
+                                image.source: root.selectionBadgeImage
```
This one needs a border in case the selected image is an identicon. Just as with the `StatusListItemBadge`, you'd get all the information you need if this was turned into `StatusImageSettings`.

alexandraB99 replied:
See https://github.com/status-im/StatusQ/pull/301#discussion_r683337557

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,198 @@
+                            anchors.centerIn: parent
+                            spacing: 2
+                            StatusBaseText {
+                                text: "In: "
```
We probably should either alias this button and expose it or introduce a property on the search popup, so that this label can be internationalized.

PascalPrecht replied:
Also this needs to use `Theme.palette.directColor1`

alexandraB99 replied:
done

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,198 @@
+                                name: root.selectionBadgePrimaryText
+                            }
+                            StatusBaseText {
+                                text: root.selectionBadgePrimaryText
```
This needs to use `Theme.palette.directColor1` as well

alexandraB99 replied:
done

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,198 @@
+                                Layout.alignment: Qt.AlignVCenter
+                                visible: !!root.selectionBadgeSecondaryText
+                                color: Theme.palette.baseColor1
+                                icon: "next"
```
The size of this icon is incorrect.
<img width="262" alt="Screenshot 2021-08-03 at 23 15 06" src="https://user-images.githubusercontent.com/445106/128087161-bf77afa7-d854-40c3-afb0-0a7c580b28a7.png">

This was designed:

<img width="290" alt="Screenshot 2021-08-03 at 23 15 26" src="https://user-images.githubusercontent.com/445106/128087212-0cfddb9a-55e5-4ffb-9dd5-d8e083e53e49.png">

alexandraB99 replied:
done. however I can't use 4.5 - 7.5 values from figma because it looks tiny, so I increased to 14.5 - 17.5

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,198 @@
+                            }
+                            StatusIcon {
+                                Layout.alignment: Qt.AlignVCenter
+                                icon: "chevron-down"
```
The size of this icon is off.

alexandraB99 replied:
done, same as above

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,198 @@
+                            StatusBaseText {
+                                text: "In: "
+                            }
+                            StatusRoundedImage {
```
In front of this image (or letter identicon) there's a possibility that we need to render another icon, in case we're selecting a private group chat type:

<img width="247" alt="Screenshot 2021-08-03 at 23 18 36" src="https://user-images.githubusercontent.com/445106/128087599-b0f75cd1-bd3b-4e19-a905-84599f75e8f4.png">

You can check  https://github.com/status-im/StatusQ/blob/master/src/StatusQ/Components/StatusChatListItem.qml#L126-L137 to see what you need for this.

alexandraB99 replied:
done, added

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,198 @@
+                                implicitWidth: 16
+                                implicitHeight: 16
+                                letterSize: 11
+                                visible: ((root.selectionBadgeImage === "") && (root.selectionBadgePrimaryText !== "Anywhere"))
```
Can't rely on `"Anywhere"` here as the text could be in a different language when i18n'ed.

Do we have other heuristics?

alexandraB99 replied:
done

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,198 @@
+                        }
+                    }
+                    onClicked: { searchPopupMenuLoader.item.popup(); }
+                }
```
As seen in previous screenshots attached in this review, the background of this button doesn't match the design. This needs update as well

alexandraB99 replied:
Done, updated. Also switched StatusBaseButton as it was (Rectangle) and made this one a QtQuick Controls 2 Button instead

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,198 @@
+                    height: 32
+                    anchors.left: searchOptionsMenuButton.right
+                    anchors.verticalCenter: parent.verticalCenter
+                    opacity: (root.selectionBadgePrimaryText === "Anywhere") ? 0.0 : 1.0
```
Same as above: Can't rely on `Anywhere` here.

alexandraB99 replied:
done

PascalPrecht commented on src/StatusQ/Popups/StatusSearchPopup.qml:
```diff
@@ -0,0 +1,198 @@
+
+                StatusBaseText {
+                    anchors.centerIn: parent
+                    text: "No results"
```
Needs to be i18n'able

alexandraB99 replied:
done

PascalPrecht reviewed (changes requested):
@alexandraB99 thank you for updating the PR!

I could now try this out and play with it and it seems to work great, aside from a few bugs I've found. Super sorry for this, don't wanna be the annoying guy, but there's a lot of UI related inconsistencies with the design as well that we need to iron out.

Also the component doesn't work well in darkmode. The comments should address all of these things.

Please let me know if anything is unclear! I've attached screenshots accordingly to make it easier to follow

alexandraB99 commented on sandbox/DemoApp.qml:
```diff
@@ -564,12 +564,150 @@ Rectangle {
+                                iconSettings.color: Theme.palette.directColor1
+                                onTriggered: {
+                                    searchPopup.resetSelectionBadge();
+                                    searchPopup.selectionBadgeImage = cryptoKittiesPopupMenu.currentImg;
```
There is a problem with StatusPopupMenu, you can try printing `currentOptionItemImgSource` for both 'Chat' and 'Cryptokitties' in DemoApp.qml  (on `Component.onCompleted`) and you'll get 
```
qml: Cryptokitties 
qml: Chat https://pbs.twimg.com/profile_images/1369221718338895873/T_5fny6o_400x400.jpg
```
so `currentOptionItemImgSource` for 'Cryptokitties' is always empty.

PascalPrecht replied:
So... the url shown in `Chat` however is the image that belongs to `CryptoKitties` in this case. Probably a off-by-one bug?

alexandraB99 replied:
Not sure why, however as this is relevant to StatusPopupMenu and not the SearchPopup, should we look fix this in this task or maybe file another bug?

PascalPrecht replied:
The currentOptionItemImgSource and other properties have been introduced in this pr so we should also fix them here, otherwise we introduce this bug in the First place just to fix it later.

alexandraB99 commented on sandbox/DemoApp.qml:
```diff
@@ -564,12 +564,150 @@ Rectangle {
-                    searchButton.onClicked: searchButton.highlighted = !searchButton.highlighted
+                    onSearchButtonClicked: {
+                        searchButton.highlighted = !searchButton.highlighted
+                        searchPopup.searchOptionsPopupMenu = searchPopupMenu;
```
not really, just for demonstration :) changed it

alexandraB99 commented on sandbox/ListItems.qml:
```diff
@@ -295,6 +295,26 @@ CExPynn1gWf9bx498P7/nzPcxEzGExhBdJGYihtAYQlO+tUZvqrPbqeudo5iJGEJjCE15a3VtodH3q2I
+        image.source: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADIAAAAyCAYAAAAeP4ixAAAAlklEQVR4nOzW0QmDQBAG4SSkl7SUQlJGCrElq9F3QdjjVhh/5nv3cFhY9vUIYQiNITSG0Bh
+CExPynn1gWf9bx498P7/nzPcxEzGExhBdJGYihtAYQlO+tUZvqrPbqeudo5iJGEJjCE15a3VtodH3q2ImYgiNITTlTdG1nUZ5a92VITQxITFiJmIIjSE0htAYQrMHAAD//+wwFVpz+yqXAAAAAElFTkSuQmCC"
+        image.isIdenticon: true
+        badge.badgeImage: "https://pbs.twimg.com/profile_images/1369221718338895873/T_5fny6o_400x400.jpg"
```
not sure I get this, could you please explain?

PascalPrecht replied:
Instead of having badeImage, it'd be `image` of type `StatusImageSettings` then you can set the `source` and other props

alexandraB99 commented on sandbox/ListItems.qml:
```diff
@@ -295,6 +295,26 @@ CExPynn1gWf9bx498P7/nzPcxEzGExhBdJGYihtAYQlO+tUZvqrPbqeudo5iJGEJjCE15a3VtodH3q2I
+        image.isIdenticon: true
+        badge.badgeImage: "https://pbs.twimg.com/profile_images/1369221718338895873/T_5fny6o_400x400.jpg"
+        badge.badgePrimaryText: "CryptoKitties"
+        badge.badgeSecondaryText: "#test"
```
sure thing, updated

alexandraB99 commented on src/StatusQ/Components/StatusListItemBadge.qml:
```diff
@@ -0,0 +1,65 @@
+                visible: !!root.badgeImage
+                image.source: root.badgeImage
+                border.color: Theme.palette.baseColor1
+            }
```
As I already wrote above: not sure if in this case the user will be able to set manually properties like `isIdenticon`. I assume this info will come from some model, does it still make sense to use StatusImageSettings?

PascalPrecht replied:
Yea so the menu is generated in status  desktop where there's access to information that sets these values

saledjenic commented:
Hey @alexandraB99 @PascalPrecht since I tried to use this PR in the main app and tried to have all that working I faced two issues which I would like you to update:

1. (minor) there is no vertical scroll bar in the search result list.
2. (blocker) the location menu (which determines where the search should be applied) must be created dynamically (using model/instantiator), currently there is no way to create content of the StatusPopupMenu component dynamically. That is a must cause that items/subitems should represent the current state in the app in terms of communities/chats/channels.

alexandraB99 commented:
@saledjenic I have added the Scrollbar and also switched to use Instantiator so that the model can be set dynamically. There are a few TODO on my side, but this should give you a heads up for now for you to work. One question: what is the thing to check on the model in order to determine whether contains a submenu or not?

@PascalPrecht I believe I have addressed all comments now. The bug found with the image is not relevant anymore, since the menu is added dynamically, I have also removed the properties. I will switch to use IconSettings in ListItemBadge tomorrow in a separate commit.

saledjenic commented:
@alexandraB99 @PascalPrecht two more things:

1. the UI should be able to render a menu based on model structure like this one:
```
ListModel {
        ListElement { text: "Chat"
            imageSource: ""
            iconName: "private-chat"
            identiconColor: ""
            isIdenticon: false
            subItems: [
                ListElement {
                    text: "Pascal"
                    imageSource: "https://pbs.twimg.com/profile_images/1369221718338895873/T_5fny6o_400x400.jpg"
                    iconName: ""
                    identiconColor: ""
                    isIdenticon: false
                },
                ListElement {
                    text: "vitalik.eth"
                    imageSource: ""
                    iconName: ""
                    identiconColor: "red"
                    isIdenticon: true
                }]
        }
}
```
so the menu based on that model should look something like:
```
|
|-> Anywhere
|-> Separator
|-> (private-chat-icon) Chat ->
                               |-> (image-source-icon) Pascal
                               |-> (V-in-a-red-circle-icon) vitalik.eth
```

2. With the current implementation we're able to select only "Anywhere" or channel/chat as a location where to search in, but not able to select whole chat section (with all containing chats) or community section (with all containing channels) as a location for the search. But that's also a location for the search, see [design](https://www.figma.com/file/Mr3rqxxgKJ2zMQ06UAKiWL/%F0%9F%92%AC-Chat%E2%8E%9CDesktop?node-id=6042%3A1).

But for that second thing maybe we need to speak to Simon as selecting whole Chat or Community section is not quite intuitive cause they are menu which contains submenus by design.

PascalPrecht reviewed (commented):
Regarding "#" in the popup menu:

Not sure if you're already doing this, but the "#" for community channel names in the popup menu should be created using `iconName: "channel"` and not be `"#"` in concatenated to the name either

PascalPrecht commented:
@saledjenic @alexandraB99 here you can see what I'm referring to in https://github.com/status-im/StatusQ/pull/301#pullrequestreview-727237770 : https://github.com/status-im/StatusQ/pull/301/commits/9fe6e0a32a9cc0864e4bfecac1f7cff920d5189b#diff-2f615fc01dbb3829146521d69af20abe6ee2e0bc519dde836e087eb2694f4be5R262

saledjenic commented:
> Regarding "#" in the popup menu:
> 
> Not sure if you're already doing this, but the "#" for community channel names in the popup menu should be created using `iconName: "channel"` and not be `"#"` in concatenated to the name either

@PascalPrecht no I didn't add it as an icon, but as a concatenated character in the model's items. Ok, will update that. Thanks.

PascalPrecht commented:
🎉 It has landed!
